File: backend/internal/model/ai.go
// Copyright (c) 2025 kk
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

package model

import (
	"time"

	"gorm.io/gorm"
)

// AIChatSession stores a user's assistant conversation thread.
type AIChatSession struct {
	ID        uint           `gorm:"primaryKey" json:"id"`
	UserID    uint           `gorm:"not null;index" json:"user_id"`
	Title     string         `gorm:"size:256" json:"title"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName specifies table name.
func (AIChatSession) TableName() string {
	return "ai_chat_sessions"
}

// AIChatMessage is one message in a chat session.
type AIChatMessage struct {
	ID        uint      `gorm:"primaryKey" json:"id"`
	SessionID uint      `gorm:"not null;index" json:"session_id"`
	Role      string    `gorm:"size:32;not null" json:"role"`
	Content   string    `gorm:"type:text;not null" json:"content"`
	CreatedAt time.Time `json:"created_at"`
}

// TableName specifies table name.
func (AIChatMessage) TableName() string {
	return "ai_chat_messages"
}

// AIAnomalyRule schedules metric anomaly checks via LLM.
type AIAnomalyRule struct {
	ID             uint           `gorm:"primaryKey" json:"id"`
	Name           string         `gorm:"size:128;not null" json:"name"`
	IntegrationID  uint           `gorm:"not null;index" json:"integration_id"`
	Query          string         `gorm:"type:text;not null" json:"query"`
	ScheduleCron   string         `gorm:"size:64;not null" json:"schedule_cron"`
	Enabled        bool           `gorm:"default:true" json:"enabled"`
	PromptTemplate string         `gorm:"type:text" json:"prompt_template"`
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName specifies table name.
func (AIAnomalyRule) TableName() string {
	return "ai_anomaly_rules"
}

// AIAnomalyFinding is one evaluated anomaly result.
type AIAnomalyFinding struct {
	ID        uint      `gorm:"primaryKey" json:"id"`
	RuleID    uint      `gorm:"not null;index" json:"rule_id"`
	Ts        time.Time `gorm:"not null;index" json:"ts"`
	Severity  string    `gorm:"size:32;not null" json:"severity"`
	Summary   string    `gorm:"type:text" json:"summary"`
	Raw       string    `gorm:"type:text" json:"raw"`
	CreatedAt time.Time `json:"created_at"`
}

// TableName specifies table name.
func (AIAnomalyFinding) TableName() string {
	return "ai_anomaly_findings"
}

// AIRcaReport stores generated root-cause analysis for an incident.
type AIRcaReport struct {
	ID          uint      `gorm:"primaryKey" json:"id"`
	IncidentID  uint      `gorm:"not null;index" json:"incident_id"`
	GeneratedAt time.Time `gorm:"not null" json:"generated_at"`
	ReportMD    string    `gorm:"type:text;not null" json:"report_md"`
	RawToolLog  string    `gorm:"type:text" json:"raw_tool_log"`
	CreatedAt   time.Time `json:"created_at"`
}

// TableName specifies table name.
func (AIRcaReport) TableName() string {
	return "ai_rca_reports"
}

File: frontend/src/api/ai.ts
// Copyright (c) 2025 kk
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

import apiClient from './client'
import type { IntegrationPublic } from './integration'

export interface AIMessage {
  role: string
  content: string
}

export interface AIChatSessionRow {
  id: number
  user_id: number
  title: string
  created_at: string
  updated_at: string
}

export interface AIChatMessageRow {
  id: number
  session_id: number
  role: string
  content: string
  created_at: string
}

export interface AIAnomalyRuleRow {
  id: number
  name: string
  integration_id: number
  query: string
  schedule_cron: string
  enabled: boolean
  prompt_template?: string
  created_at: string
  updated_at: string
}

export interface AIAnomalyFindingRow {
  id: number
  rule_id: number
  ts: string
  severity: string
  summary: string
  raw?: string
  created_at: string
}

export interface AIRcaReportRow {
  id: number
  incident_id: number
  generated_at: string
  report_md: string
  raw_tool_log?: string
}

export const aiApi = {
  listProviders: () => apiClient.get<{ data: IntegrationPublic[] }>('/ai/providers'),

  listSessions: () => apiClient.get<{ data: AIChatSessionRow[] }>('/ai/sessions'),

  getSession: (id: number) =>
    apiClient.get<{ data: { session: AIChatSessionRow; messages: AIChatMessageRow[] } }>(
      `/ai/sessions/${id}`
    ),

  deleteSession: (id: number) => apiClient.delete(`/ai/sessions/${id}`),

  listAnomalyRules: () => apiClient.get<{ data: AIAnomalyRuleRow[] }>('/ai/anomaly/rules'),

  createAnomalyRule: (body: Partial<AIAnomalyRuleRow>) =>
    apiClient.post<{ data: AIAnomalyRuleRow }>('/ai/anomaly/rules', body),

  updateAnomalyRule: (id: number, body: Partial<AIAnomalyRuleRow>) =>
    apiClient.put<{ data: AIAnomalyRuleRow }>(`/ai/anomaly/rules/${id}`, body),

  deleteAnomalyRule: (id: number) => apiClient.delete(`/ai/anomaly/rules/${id}`),

  listFindings: (params?: { rule_id?: number; since?: string; limit?: number }) =>
    apiClient.get<{ data: AIAnomalyFindingRow[] }>('/ai/anomaly/findings', { params }),

  postRca: (incidentId: number, integrationIdOverrides?: number) =>
    apiClient.post<{
      data: { report_md: string; cited_tool_calls: unknown[]; stored_id: number }
    }>('/ai/rca', {
      incident_id: incidentId,
      integration_id_overrides: integrationIdOverrides,
    }),

  listRcaReports: (params?: { incident_id?: number; limit?: number }) =>
    apiClient.get<{ data: AIRcaReportRow[] }>('/ai/rca/reports', { params }),
}

/** Streams POST /ai/chat; invokes onToken for each chunk and returns session id from header. */
export async function streamAIChat(
  body: {
    session_id?: number
    integration_id?: number
    messages: AIMessage[]
    context?: Record<string, unknown>
  },
  onToken: (t: string) => void,
  signal?: AbortSignal
): Promise<{ sessionId: number | null }> {
  const token = localStorage.getItem('token')
  const res = await fetch('/api/v1/ai/chat', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(body),
    signal,
  })
  const sidHdr = res.headers.get('X-Session-ID')
  const sessionId = sidHdr ? parseInt(sidHdr, 10) : null
  if (!res.ok) {
    const errText = await res.text()
    let msg = errText || res.statusText
    try {
      const j = JSON.parse(errText) as { error?: string }
      if (typeof j?.error === 'string') {
        msg = j.error
      }
    } catch {
      /* leave msg as raw body */
    }
    throw new Error(msg)
  }
  const reader = res.body?.getReader()
  if (!reader) {
    return { sessionId }
  }
  const dec = new TextDecoder()
  let buf = ''
  for (;;) {
    const { done, value } = await reader.read()
    if (done) break
    buf += dec.decode(value, { stream: true })
    const parts = buf.split('\n')
    buf = parts.pop() ?? ''
    for (const line of parts) {
      if (!line.startsWith('data: ')) continue
      try {
        const data = JSON.parse(line.slice(6)) as { token?: string; done?: boolean }
        if (data.token) onToken(data.token)
      } catch {
        /* ignore malformed SSE lines */
      }
    }
  }
  return { sessionId }
}
